graphql-java-support/src/main/java/com/apollographql/federation/graphqljava/tracing/FederatedTracingInstrumentation.java

```diff
@@ -314,6 +291,138 @@ private Timestamp getNowTimestamp() {
         private long getDuration() {
             return System.nanoTime() - startRequestNanos;
         }
+
+        /**
+         * A thread-safe class for storing field information and converting it to a protobuf.
+         */
+        private static class ProtoBuilderTree {
+            private final Node root;
+            // It's necessary to use ConcurrentHashMap here due to its computeIfAbsent guaranteeing
+            // at-most-once execution of the lambda (ConcurrentMap does not guarantee this).
+            private final ConcurrentHashMap<ExecutionPath, Node> nodesByPath;
+            // We use a whole-tree read-write lock to prevent the protobuf conversion step from
+            // having to acquire each node's lock.
+            private final ReadWriteLock treeLock;
+            // This flag is used to ensure protobuf conversion occurs once, and that builders can't
+            // be edited afterwards.
+            private boolean isFinalized;
+
+            public ProtoBuilderTree() {
+                root = new Node(Reports.Trace.Node.newBuilder());
+                nodesByPath = new ConcurrentHashMap<>();
+                nodesByPath.put(ExecutionPath.rootPath(), root);
+                treeLock = new ReentrantReadWriteLock();
```

Comment:
Why is reentrancy necessary?

Reply:
Reentrance isn't necessary here for correctness. It was mainly to support any future changes where we:
1) add a public function that would need to acquire a lock
2) call that public function from another function that may already have acquired the lock

Currently public functions don't call other public functions, so we don't need reentrancy here. (Just documentation for those functions about what locks are needed.)

Instead of preemptively making the lock reentrant, I'll use a non-reentrant version, and we can shift to reentrant in the future if the need arises.

Reply:
Actually it turns out that the only implementing class of the `ReadWriteLock` interface is `ReentrantReadWriteLock`, so I guess we're stuck with it 🤷‍♂.